core/src/main/java/nl/nn/adapterframework/stream/Message.java

```diff
@@ -495,120 +507,141 @@ public long size() {
-						if (log.isTraceEnabled()) log.trace("FilterInputStream.read");
-						int result = super.read(b, off, len);
-						if (result>=0 && pos<maxSize) {
-							pos+=result;
-							stream.write(b, off, result);
-							if (log.isTraceEnabled()) log.trace("captured ["+result+"] bytes");
-						}
-						return result;
+					return result;
+				}
+
+				@Override
+				public int read(byte[] b, int off, int len) throws IOException {
+					if (log.isTraceEnabled()) log.trace("FilterInputStream.read");
+					int result = super.read(b, off, len);
+					if (result>=0 && pos<maxSize) {
+						pos+=result;
+						outputStream.write(b, off, result);
+						if (log.isTraceEnabled()) log.trace("captured ["+result+"] bytes");
 					}
-
-					@Override
-					public void close() throws IOException {
-						if (log.isTraceEnabled()) log.trace("FilterInputStream.close");
-						if (available()>0) {
-							if (log.isTraceEnabled()) log.trace("bytes available at close");
-							stream.write("(--MoreBytesAvailable--)[".getBytes());
-							read(new byte[1000]);
-							stream.write("]".getBytes());
-							if (read()>0) {
-								stream.write("...".getBytes());
-							}
-						}
-						super.close();
-						stream.close();
+					return result;
+				}
+
+				@Override
+				public void close() throws IOException {
+					if (log.isTraceEnabled()) log.trace("FilterInputStream.close");
+					if (available()>0) {
+						// Make the bytes available for debugger even when the stream was not used (might be because the
+						// pipe or sender that normally consumes the stream is stubbed by the debugger)
+						int len = read(new byte[maxSize]);
+						if (log.isTraceEnabled()) log.trace(len+" bytes available at close");
 					}
+					super.close();
+					outputStream.close();
+				}
 
-				};
-			} catch (IOException e) {
-				log.warn("Cannot capture stream", e);
-				return null;
-			}
-		} else {
-			try {
-				request = new FilterReader(asReader()) {
-
-					int pos;
-					
-					@Override
-					public int read() throws IOException {
-						if (log.isTraceEnabled()) log.trace("FilterReader.read");
-						int result = super.read();
-						if (result>=0 && pos++<maxSize) {
-							writer.write((char)result);
-						}
-						return result;
+			};
+		} catch (IOException e) {
+			log.warn("Cannot capture stream", e);
+			return null;
+		}
+		return outputStream;
+	}
+	
+	/**
+	 * Can be called when {@link #requiresStream()} it true and {@link #isBinary()} is false to retrieve a copy of (part
+	 * of) the character stream that is in this message, after the stream has been closed.
+	 */
+	public StringWriter captureCharacterStream() {
+		return captureCharacterStream(new StringWriter());
+	}
+	public <W extends Writer> W captureCharacterStream(W writer) {
+		return captureCharacterStream(10000, writer);
+	}
+	public <W extends Writer> W captureCharacterStream(int maxSize, W writer) {
+		if (!requiresStream() || isBinary()) {
```

Comment:
De test op isBinary() hoeft hier in principe niet.

Reply:
Ah, dat is antwoord op een vraag die ik net heb gesteld (had dit nog niet gelezen). :)